src/main/java/org/broadinstitute/hellbender/utils/Utils.java

```diff
@@ -1041,22 +1048,336 @@ public static boolean xor(final boolean x, final boolean y) {
+    public static ImmutablePair<Integer, Integer> oneIndelHapToRef(final byte[] reference, final byte[] query, int maxIndelLength){
+        int lengthOfIndel = Math.abs(reference.length - query.length);
+
+        if(lengthOfIndel > maxIndelLength){
+            return new ImmutablePair<Integer, Integer>(-1,0);
+        }
+        //deletion
+        if(query.length < reference.length){
+            return findIndel(query, reference, lengthOfIndel);
+        }
+        //insertion
+        else if(query.length > reference.length){
+            return findIndel(reference, query, lengthOfIndel);
+        }
+        //lengths are equal, no one indel
+        else{
+            return new ImmutablePair<Integer, Integer>(-1,0);
+        }
+    }
+
+    private static ImmutablePair<Integer, Integer> findIndel(final byte[] shorter, final byte[] longer, int lengthOfIndel){
+        int indelEnd = -1;
+        //traverse backwards until you hit mismatch/end of indel
+        for(int i = shorter.length - 1; i >= 0; i--){
+            if(shorter[i] != longer[i + lengthOfIndel]){
+                indelEnd = i + 1;
+                break;
+            }
+        }
+        if(indelEnd == -1){
+            //deletion located at beginning of the query
+            return new ImmutablePair<Integer, Integer>(0, lengthOfIndel);
+        }
+        //traverse forwards until you hit start of indel
+        for(int i = 0; i < indelEnd; i++){
+            if(shorter[i] != longer[i]){
+                return new ImmutablePair<Integer, Integer>(-1,0);
+            }
+        }
+        //traversed entire query, one indel and no mismatches found
+        return new ImmutablePair<Integer, Integer>(indelEnd, lengthOfIndel);
+    }
+
+
+    /**
+     * Finds the location of one indel in the query sequence in relation to the reference sequence
+     * local Alignment
+     *
+     * Returns an Indel object that contains all the necessary information for generating a cigar string
+     *
+     * @param reference the reference sequence
+     * @param query the query sequence
+     * @param parameters the SW parameters
+     * @param maxInsertionSize allowed size of insertion so that we can confidently say nothing can beat it
+     * @param maxDeletionSize allowed size of deletion so that we can confidently say nothing can beat it
+     */
+    public static Indel oneIndelReadToHap(final byte[] reference, final byte[] query, SWParameters parameters, int maxInsertionSize, int maxDeletionSize){
+
+        Indel insertion = new Indel(-1, -1, maxInsertionSize, true);
+        Indel deletion = new Indel(-1, -1, maxDeletionSize, false);
+
+        //set bounds for indices that are eligible to have indels
+        int refBackInsertionBound = query.length - 1 - maxInsertionSize;
+        int refBackDeletionBound = query.length;
+
+        int refIndexBack = reference.length - 1;
+
+        //traverse ref up to insertion bound
+        while(refIndexBack >= refBackInsertionBound){
```

Comment:
I would make this a for loop, i don't think skipping ~3/4 of start sites with the queue of start positions with the same starting base is worth the complexity that comes from having to reset refIndexBack. Just iterate from the back with refIndexBack and then use another integer for local operations.